step-ca/templates.mdx

```diff
@@ -150,7 +151,111 @@ You can also import values
 from the `"templateData"` object
 in your provisioner's [configuration block](#configuring-the-ca-to-use-templates).
 
-### SSH Templates
+### ASN.1 Values
+
+The X.509 templating system includes a few functions for encoding ASN.1 values.
+Use these functions to populate custom OID `extensions`:
+
+```
+"extensions": [
+  {"id": "1.2.3.4", "critical": false, "value": {{ asn1Enc "int:3848281" | toJson }}}
+]
+```
+
+When applied to template variables, these functions enable dynamic OID extensions:
+
+```
+{
+	"subject": {{ toJson .Subject }},
+	"sans": {{ toJson .SANs }},
+{{- if typeIs "*rsa.PublicKey" .Insecure.CR.PublicKey }}
+	"keyUsage": ["keyEncipherment", "digitalSignature"],
+{{- else }}
+	"keyUsage": ["digitalSignature"],
+{{- end }}
+	"extKeyUsage": ["serverAuth", "clientAuth"],
+	"extensions": [
+		{"id": "1.2.3.4", "value": {{ asn1Marshal .AuthorizationCrt.NotAfter | toJson }}},
+		{"id": "1.2.3.5", "value": {{ asn1Set (asn1Marshal (first .Insecure.CR.DNSNames) "utf8") (asn1Enc "int:123456") | toJson }}},
+		{"id": "1.2.3.6", "value": {{ asn1Seq (asn1Enc "YubiKey") (asn1Enc "int:123456") | toJson }}}
+	]
+}
+```
+
+Let's walk through one line of this template:
+
+```
+{"id": "1.2.3.6", "value": {{ asn1Seq (asn1Enc "YubiKey") (asn1Enc "int:123456") | toJson }}}
+```
+
+To render this line of the template:
+
+1. `asn1Enc "YubiKey"` encodes "YubiKey" as an ASN.1 Printable string, and returns the Base64
+2. `asn1Enc "int:123456"` encodes 123 as an ASN.1 Integer, and returns the Base64
+3. `asn1Seq` gets the resulting Base64 strings, decodes them, creates an ASN.1 SEQENCE, and returns the Base64 of the SEQUENCE.
+4. Finally, `toJson` returns a JSON string representation of the Base64
+
+#### asn1Enc
+
+Encodes a string into an ASN.1 value.
+
+A data type can be supplied as a prefix,
+and ASN.1 Printable is the default data type:
+
+```
+asn1Enc "int:123"
+asn1Enc "oid:1.2.3.4"
+asn1Enc "foo"
+asn1Enc .Subject.CommonName
+```
+
+This function supports the data types described [here](https://github.com/smallstep/crypto/blob/878092791d7e5ef8530dfa949cfe3cde61373d77/x509util/extensions.go#L486):
+
+- **printable**: "printable:abcfoo" (or just "abcfoo")
+- **int**: "int:12324"
+- **oid**: "oid:1.2.3.4"
+- **utf8**: "utf8:foobar"
+- **ia5**: "ia5:abcfoo"
+- **numeric**: "numeric:01 02 03345" (multiple space-separated ints)
```

Comment:
It allows numbers and spaces, you can also use numeric:1234, you don't have to use spaces.